src/utils/net/http_client.ml

```diff
@@ -89,440 +91,246 @@ let basic_request = {
+
+    (match r.req_request with
+    | GET -> ()
+    | HEAD ->
+      Curl.set_nobody curl true;
+      Curl.set_header curl true;
+    );
+
+    (match r.req_referer with
+    | Some ref_url -> Curl.set_referer curl (Url.to_string_no_args ref_url)
+    | None -> ());
+
+    (* TODO: TEST *)
+    (match r.req_proxy with
+    | Some (host, port, Some (username, password)) ->
+        Curl.set_proxy curl host;
+        Curl.set_proxyport curl port;
+        Curl.set_proxyuserpwd curl (username ^ ":" ^ password)
+    | Some (host, port, None) ->
+        Curl.set_proxy curl host;
+        Curl.set_proxyport curl port
+    | None -> ());
+
+    (* TODO: TEST *)
+    if (int_of_float r.req_max_total_time) <> (int_of_float infinite_timeout) then
+      Curl.set_timeout curl (int_of_float r.req_max_total_time);
+
+    Curl.set_writefunction curl write_f;
+    Curl.perform curl;
+    let http_code = Curl.getinfo curl Curl.CURLINFO_HTTP_CODE in
+    Curl.cleanup curl;
+    match http_code with
+    | Curl.CURLINFO_Long code -> (match code with
+      | 200 ->
+        lprintf_nl "HTTP success: %s" (Url.to_string r.req_url);
+        Ok ""
+      | 400 when r.req_request = HEAD ->
         lprintf_nl "Error 400 received for HEAD %s, re-try GET" (Url.to_string_no_args r.req_url);
         let r2 = {
           r with
           req_request = GET;
+          req_retry = r.req_retry + 1;
         } in
-        get_page r2 content_handler f ferr
-
-    | 404 ->
-        lprintf_nl "404: Not found for: %s" (Url.to_string_no_args r.req_url);
-        close sock (Closed_for_error "bad reply");
-        ferr (`HTTP ans_code);
-        raise Not_found
-
-    | 502 | 503 | 504 ->
-        if !verbose then lprintf_nl "%d: Unavailable" ans_code;
-  let retrynum = r.req_retry in
-        if retrynum < r.req_max_retry then begin
-          if !verbose then print_headers ();
-      let seconds = (retrynum+1)*10 in
-              lprintf_nl "retry %d/%d in %d seconds for %s"
-          (retrynum+1) r.req_max_retry seconds (Url.to_string_no_args r.req_url);
-          let r = { r with 
-            req_retry = retrynum+1 
-          } in
-          add_timer (float(seconds)) (fun t -> get_page r content_handler f ferr)
-    end
-        else begin
-          lprintf_nl "more than %d retries, aborting." r.req_max_retry;
-          ferr (`HTTP ans_code);
-          raise Not_found
-        end
-          
+        http_call_internal r2 write_f fretry progress
+      | 502 | 503 | 504 ->
+        raise (ScheduleRetry (`HTTP code))
+      | code ->
+        lprintf_nl "HTTP error occurred: %d %s" code (Url.to_string r.req_url);
+        Error (`HTTP code))
     | _ ->
-        lprintf_nl "%d: bad reply for: %s"
-          ans_code (Url.to_string_no_args r.req_url);
-        close sock (Closed_for_error "bad reply");
-        ferr (`HTTP ans_code);
-        raise Not_found
+      lprintf_nl "HTTP error unknown: %s" (Url.to_string r.req_url);
+      Curl.cleanup curl;
```

Comment:
duplicate cleanup
better wrap it with `bracket` (copy paste from tools/bt_dht_node.ml) (there is `Fun.protect` in stdlib but not in older ocaml versions)

Reply:
Is the newer version what you were thinking?